use tensoriterator instead of th for fill_ implementation.

nairbv opened the pull request:
Moves fill_ to aten as suggested in:
https://github.com/pytorch/pytorch/pull/20336#issuecomment-493260729

borrows from @cpuhrsch's PR: https://github.com/pytorch/pytorch/pull/18876/files#diff-0d1178f1a4ce15aeb760d251974e6924

Co-authored-by: @cpuhrsch

nairbv committed:
use tensoriterator instead of th for fill_ implementation.

nairbv committed:
fix tests broken by moving fill to aten

fmassa commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -85,12 +85,18 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
 }
 
+Tensor& fill_out(Tensor& self, const Scalar value) {
+  auto iter = TensorIterator::unary_op(self, self);
```
Question: should this be a nullary op instead?

nairbv replied:
I wasn't very confident and I'm pretty open to the idea that nullary might be preferable. We don't currently have any `NullaryOpsKernel` etc.

Unary seems to make sense semantically, in a file with operations like abs() etc.

The macros that provide TH implementations for some of the other unary ops also seems somewhat similar with the (self, self) for in-place operations, e.g.:
```
  Tensor& _##op##__cpu(Tensor& self) {                          \
    return at::op##_out(self, self);                            \
  }                                                             \
```

Another option is that it could be clearer as unary if the second arg is a Tensor, called by the Scalar implementation by using wrapped_scalar_tensor(scalar_value) as is done in many of the BinaryOps. I felt like if I converted to Tensor I'd just need to get back to a Scalar anyway though, so didn't take that path.

colesbury replied:
You should use `TensorIterator::nullary_op` since this only writes self and doesn't read self. It's fine to keep the definition in this file -- there's a reasonable semantic argument that this is a "unary" operation even if TensorIterator doesn't care about `Scalar value`.

I agree that it's probably best not to use wrapped_scalar_tensor here.

This will require implementing a few new functions in Loops.h: `nullary_kernel_vec`, `nullary_kernel`, and their transitive dependencies. You should be able to grab them from @cpuhrsch's PR: https://github.com/pytorch/pytorch/pull/18876/files#diff-0d1178f1a4ce15aeb760d251974e6924

On the GPU side, there is already a `gpu_nullary_kernel`.

nairbv replied:
I was thinking of it as total inputs to create the result (need a value to fill with), which is why I was thinking of it as unary, but, will also be easier to implement with nullary now that I can grab the loop from that PR. Thanks.

nairbv replied:
looking more closely at the linked PR, I didn't realize someone was already working on fill_. I was looking at it because of this PR: https://github.com/pytorch/pytorch/pull/20336#issuecomment-493260729 and also just wanted to understand TensorIterator better.

nairbv committed:
git add unadded cuda file

colesbury commented:
FYI, @cpuhrsch has an unfinished PR here https://github.com/pytorch/pytorch/pull/18876

colesbury commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -54,6 +54,25 @@ static void abs_kernel(TensorIterator& iter) {
 
+static void fill_kernel(TensorIterator& iter, Scalar value_scalar) {
+  if( iter.dtype() == ScalarType::Half ) {
+    auto value = value_scalar.to<at::Half>().x;
```
Clever. We should probably do this in copy_kernel as well.

nairbv replied:
We might want to rename `x`. Won't work in cases where we want to do any math on the underlying data, but yeah, should work for copy/fill.

nairbv committed:
Fix a test and switch unary->nullary op

copied some nullary/loops code from:
https://github.com/pytorch/pytorch/pull/18876

Co-authored-by: cpuhrsch

colesbury commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -85,12 +85,19 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
 
 Tensor& fill_(Tensor& self, const Tensor& value) {
-  return at::legacy::th::_th_fill_(self, value);
+  AT_ASSERTM(value.dim() == 0, "fill_ only supports 0-dimension value tensor.");
```
use `TORCH_CHECK`

The previous error message includes the dimension of value which is sometimes helpful for debugging:

"fill only supports a 0-dimensional value tensor, but got tensor with 1 dimension(s)"

colesbury commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -216,5 +236,6 @@ IMPLEMENT_FLOAT_KERNEL(FLOATING, sqrt)
 IMPLEMENT_FLOAT_KERNEL(FLOATING, tan)
 IMPLEMENT_FLOAT_KERNEL(FLOATING, tanh)
 IMPLEMENT_FLOAT_KERNEL(FLOATING, trunc)
+//IMPLEMENT_FLOAT_KERNEL(FLOATING, fill)
```
remove

colesbury reviewed (approved):
lgtm (but see inline comments)

cpuhrsch commented:
Feel free to recycle https://github.com/pytorch/pytorch/pull/18876. I'll close it after this merges.

nairbv committed:
small improvements to fill_ implementation
